Schema Viewer Drawer

emtwo opened the pull request:
This is a fresh PR with the code from https://github.com/getredash/redash/pull/2990 rebased and linted.

It is ready for review now. This PR is the first of a series of PRs for schema enhancements. I will link the subsequent PRs here as they become available.

[1] Schema viewer drawer https://github.com/getredash/redash/pull/3291 (this one)
[2] Schema admin configuration https://github.com/getredash/redash/pull/3292
[3] Schema query samples https://github.com/getredash/redash/pull/3293
[4] Data source descriptions #3401

emtwo commented:
Note: the way that schema updates work now is through a periodic celery task that runs the queries to get column names and types etc. The results are stored in the new schema tables. Whenever the schema is fetched from the UI, it just directly queries the data in these tables.

Since the schema is set to refresh only every 30 min (https://github.com/getredash/redash/blob/master/redash/settings/__init__.py#L48), this is likely why the `percy/redash` visual error shows up.

We can either increase the frequency of schema update (quicker option,  but not as good) or have a one-off schema refresh that is done on init so that the schema is available. I'll look into the latter.

gabrieldutra commented on cypress/integration/query/create_query_spec.js:
```diff
@@ -1,13 +1,33 @@
+    };
+
     cy.login();
+    cy.request({
```
I've created [db-seed.js](https://github.com/getredash/redash/blob/master/cypress/seed-data.js) with this purpose :thinking:, so this kind of dependency could be created by using `npm run cypress db-seed`, lmk what you think haha

emtwo replied:
@gabrieldutra I was in fact, just testing. Though I could use some help. I cannot reproduce this percy issue locally that shows up here. In fact, when I run the `create_query_spec.js` test on `master` locally, the DOM snapshots seem to be missing the shema data (included screenshot below) And on the other hand, the snapshot for this PR seems to show the schema locally.

Any idea what might be going on here?

<img width="1387" alt="screen shot 2019-01-21 at 5 18 28 pm" src="https://user-images.githubusercontent.com/784781/51501647-cb0b5080-1da0-11e9-84b0-0187c1b805f6.png">

gabrieldutra replied:
Also, I don't know if it's related, but I noticed the **Chinook** data source is not showing schema info in the [preview](https://deploy-preview-3291--redash-preview.netlify.com/queries/new).

I'll try to reproduce this locally and give you some help with Percy anyway

arikfr replied:
> I have faced some issues when I was doing changes in frontend and handling Cypress in development mode. The frontend code seems not to be shared within the docker container.

Does the Cypress Docker Compose configuration use VOLUMEs?

gabrieldutra replied:
Cypress is using `docker-compose.cypress.yml` when in CI and the development `docker-compose.yml` when not.

However it uses `http://localhost:5000`, which I guess doesn't use webpack to watch files, so frontend in this case only updates after a rebuild. The two options I see to make it friendlier to the developer would be either adding a `npm run start` to a frontend container in `docker-compose.yml` or adding this outside docker in cypress scripts.

arikfr replied:
I think we should add to the instructions to run `npm run build` before running Cypress tests. Running npm in the container is not possible, because the container will not have Node (currently it does, but it's a temporary thing).

emtwo commented on docker-compose.yml:
```diff
@@ -29,7 +29,7 @@ services:
       REDASH_REDIS_URL: "redis://redis:6379/0"
       REDASH_DATABASE_URL: "postgresql://postgres@postgres/postgres"
-      QUEUES: "queries,scheduled_queries,celery"
+      QUEUES: "queries,scheduled_queries,celery,schemas"
```
Aweosme! Updating `docker-compose.cypress.yml` did the trick! I didn't realize cypress had its own yml file. Thank you for your help @gabrieldutra!

gabrieldutra replied:
You're welcome @emtwo! :slightly_smiling_face:

emtwo commented:
I've rebased the PR again and the original percy issue is fixed. Note that currently percy is failing for an expected reason - there are 2 new tables added - `column_metadata` and `table_metadata` that show up in the schema viewer.

gabrieldutra commented:
> I've rebased the PR again and the original percy issue is fixed. Note that currently percy is failing for an expected reason - there are 2 new tables added - `column_metadata` and `table_metadata` that show up in the schema viewer.

Don't forget to add the `schema` queue in the other files (such as `docker-compose.production.yml`) as this could cause some bugs in the future :grin:

emtwo commented:
> > I've rebased the PR again and the original percy issue is fixed. Note that currently percy is failing for an expected reason - there are 2 new tables added - `column_metadata` and `table_metadata` that show up in the schema viewer.
> 
> Don't forget to add the `schema` queue in the other files (such as `docker-compose.production.yml`) as this could cause some bugs in the future 😁

I've added the schemas queue in a couple of other spots as you suggested. However, I was hesitant at first to add it since the `schemas` queue was used in the code prior to this PR but it did not appear in any of the docker files. So without knowing the use cases of all the docker files, it's hard to tell if it's required in them. Though, I'm sure it doesn't hurt to add them. Perhaps @arikfr might have more insight into this.

arikfr commented:
I will do a review of all the Docker Compose files and add `schemas` where needed.

I do realize now that everyone who are using the AMIs we build, use a Docker Compose setup without this queue. Which means that: 1) this queue is growing in size, but nothing is processing it; 2) they don't get schema refreshes. 🤦‍♂️

arikfr commented:
Change of plans: #3325.

washort commented on redash/models/__init__.py:
```diff
@@ -198,21 +198,25 @@ def delete(self):
         return res
 
     def get_schema(self, refresh=False):
-        key = "data_source:schema:{}".format(self.id)
```
Is removing this redis caching of schema information intentional? Is there a performance impact?

emtwo replied:
Thanks for pointing this out @washort!

It was intentional because from what I recall back in the Berlin work-week, I think @arikfr  was saying he felt that using redis to store schema was a bit of a hack and he would prefer it stored in a table. Of course, we could be storing the data in tables and have additional caching for performance, but I felt this added complexity of maintaining both a cache and tables for the same data was perhaps not worth the performance gain.

I did a quick test on my machine and with 5 runs of the old vs. the new `get_schema()` function, the redis one averages 7.2ms per call and this one (from this pr) averages 44ms per call. It's a big relative difference, but 44ms isn't so bad. Though of course this could be worse in different scenarios, e.g. slower network/machine or more data. I suppose I will defer this decision to @arikfr

washort replied:
Great, just curious.

washort commented on redash/models/__init__.py:
```diff
@@ -67,6 +67,56 @@ def get(self, query_id):
 
 
+@python_2_unicode_compatible
+class TableMetadata(db.Model):
```
you'll need a migration to create these tables

emtwo replied:
Ah! I had missed this, thank you!

jezdez commented on redash/models/__init__.py:
```diff
@@ -137,28 +187,37 @@ def get_by_id(cls, _id):
+            } for column in columns], key=lambda column: column['name'])
+            schema.append(table_info)
+
+        return sorted(schema, key=lambda table: table['name'])
```
There is a more canonical way to do this:

```python
from operator import itemgetter
# ...
return sorted(schema, key=itemgetter('name'))
```

jezdez commented on redash/models/__init__.py:
```diff
@@ -137,28 +187,37 @@ def get_by_id(cls, _id):
+                'type': column.column_type,
+                'exists': column.column_exists,
+                'example': column.column_example
+            } for column in columns], key=lambda column: column['name'])
```
There is a more canonical way to do this:

```python
from operator import itemgetter
# ...
            } for column in columns],
            key=itemgetter('name'))
```

jezdez commented on redash/query_runner/__init__.py:
```diff
@@ -118,6 +119,23 @@ def _run_query_internal(self, query):
+
+        results, error = self.run_query(query, None)
+        if error is not None:
+            raise Exception("Failed getting table sample.")
```
I wonder if this should raise `NotSupported`?

jezdez commented on redash/query_runner/presto.py:
```diff
@@ -72,25 +73,30 @@ def type(cls):
+        table_samples = {}
 
-        for row in results['rows']:
+        for i, row in enumerate(results['rows']):
```
I think the `enumerate` isn't needed here anymore?

jezdez commented on redash/tasks/queries.py:
```diff
@@ -379,7 +380,64 @@ def refresh_schema(data_source_id):
+            persisted_table = models.db.session.query(
+                TableMetadata).filter(
+                TableMetadata.table_name==table_name).filter(
+                TableMetadata.data_source_id==ds.id).first()
```
Let's use this here:

```
persisted_table = TableMetadata.query.filter(
    TableMetadata.table_name == table_name,
    TableMetadata.data_source_id == ds.id,
).first()
```

jezdez commented on redash/tasks/queries.py:
```diff
@@ -379,7 +380,64 @@ def refresh_schema(data_source_id):
+                TableMetadata.data_source_id==ds.id).first()
+
+            if persisted_table:
+                models.db.session.query(TableMetadata).filter(TableMetadata.id==persisted_table.id).update({"table_exists": True})
```
```suggestion
                TableMetadata.query.filter(
                    TableMetadata.id == persisted_table.id,
                ).update({"table_exists": True})
```

jezdez commented on redash/tasks/queries.py:
```diff
@@ -379,7 +380,64 @@ def refresh_schema(data_source_id):
+                persisted_column = models.db.session.query(ColumnMetadata.id).filter(
+                    ColumnMetadata.column_name==column, ColumnMetadata.table_id==persisted_table.id).first()
+                if persisted_column:
+                    models.db.session.query(ColumnMetadata).filter(ColumnMetadata.id==persisted_column.id).update(column_metadata)
```
```suggestion
                    ColumnMetadata.query.filter(
                        ColumnMetadata.id == persisted_column.id,
                    ).update(column_metadata)
```

jezdez commented on redash/tasks/queries.py:
```diff
@@ -379,7 +380,64 @@ def refresh_schema(data_source_id):
+
+                # If the column exists, update it, otherwise create a new one.
+                persisted_column = models.db.session.query(ColumnMetadata.id).filter(
+                    ColumnMetadata.column_name==column, ColumnMetadata.table_id==persisted_table.id).first()
```
```python
                 persisted_column = ColumnMetadata.query.filter(
                     ColumnMetadata.column_name == column,
                     ColumnMetadata.table_id == persisted_table.id,
                 ).options(load_only('id')).first()
```

jezdez commented on redash/tasks/queries.py:
```diff
@@ -379,7 +380,64 @@ def refresh_schema(data_source_id):
+
+            # If a column did not exist, set the 'column_exists' flag to false.
+            models.db.session.query(ColumnMetadata).filter(ColumnMetadata.table_id==persisted_table.id,
+                ~ColumnMetadata.column_name.in_(tuple(existing_columns))).update({"column_exists": False}, synchronize_session='fetch')
```
```python
             existing_columns_list = list(existing_columns)
             ColumnMetadata.query.filter(
                 ColumnMetadata.table_id == persisted_table.id,
                 ~ColumnMetadata.column_name.in_(existing_columns_list),
             ).update({
                 "column_exists": False
             }, synchronize_session='fetch')
```

jezdez commented on redash/tasks/queries.py:
```diff
@@ -379,7 +380,64 @@ def refresh_schema(data_source_id):
+        tables_to_update = models.db.session.query(TableMetadata).filter(
+            ~TableMetadata.table_name.in_(tuple(existing_tables)))
+
+        tables_to_update.update({"table_exists": False}, synchronize_session='fetch')
```
```python
         tables_to_update = TableMetadata.query.filter(
             ~TableMetadata.table_name.in_(existing_columns_list)
         ).update({"table_exists": False}, synchronize_session='fetch')
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+        }
+
+        refresh_schema(self.factory.data_source.id)
+        table_metadata = models.db.session.query(TableMetadata).all()
```
```suggestion
        table_metadata = TableMetadata.query.all()
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+
+        refresh_schema(self.factory.data_source.id)
+        table_metadata = models.db.session.query(TableMetadata).all()
+        column_metadata = models.db.session.query(ColumnMetadata).all()
```
```suggestion
        column_metadata = ColumnMetadata.query.all()
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+
+    def test_refresh_schema_deleted_table_marked(self):
+        refresh_schema(self.factory.data_source.id)
+        table_metadata = models.db.session.query(TableMetadata).all()
```
```suggestion
        table_metadata = TableMetadata.query.all()
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+    def test_refresh_schema_deleted_table_marked(self):
+        refresh_schema(self.factory.data_source.id)
+        table_metadata = models.db.session.query(TableMetadata).all()
+        column_metadata = models.db.session.query(ColumnMetadata).all()
```
```suggestion
        column_metadata = ColumnMetadata.query.all()
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+        self.patched_get_schema.return_value = []
+
+        refresh_schema(self.factory.data_source.id)
+        table_metadata = models.db.session.query(TableMetadata).all()
```
```suggestion
        table_metadata = TableMetadata.query.all()
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+
+        refresh_schema(self.factory.data_source.id)
+        table_metadata = models.db.session.query(TableMetadata).all()
+        column_metadata = models.db.session.query(ColumnMetadata).all()
```
```suggestion
        column_metadata = ColumnMetadata.query.all()
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+    def test_refresh_schema_delete_column(self):
+        NEW_COLUMN_NAME = 'new_column'
+        refresh_schema(self.factory.data_source.id)
+        column_metadata = models.db.session.query(ColumnMetadata).all()
```
```suggestion
        column_metadata = ColumnMetadata.query.all()
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+        # Table is back, `table_exists` should be True again.
+        self.patched_get_schema.return_value = self.default_schema_return_value
+        refresh_schema(self.factory.data_source.id)
+        table_metadata = models.db.session.query(TableMetadata).all()
```
```suggestion
        table_metadata = TableMetadata.query.all()
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+        }]
+
+        refresh_schema(self.factory.data_source.id)
+        column_metadata = models.db.session.query(ColumnMetadata).all()
```
```suggestion
        column_metadata = ColumnMetadata.query.all()
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+        UPDATED_COLUMN_TYPE = 'varchar'
+
+        refresh_schema(self.factory.data_source.id)
+        column_metadata = models.db.session.query(ColumnMetadata).all()
```
```suggestion
        column_metadata = ColumnMetadata.query.all()
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+        self.patched_get_schema.return_value = updated_schema
+
+        refresh_schema(self.factory.data_source.id)
+        column_metadata = models.db.session.query(ColumnMetadata).all()
```
```suggestion
        column_metadata = ColumnMetadata.query.all()
```

jezdez commented on migrations/versions/280daa582976_.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Would you mind removing this comment and reformatting the file for pep8?

jezdez commented on migrations/versions/280daa582976_.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Same here

jezdez commented on redash/models/__init__.py:
```diff
@@ -63,6 +63,56 @@ def get(self, query_id):
 
 
+@python_2_unicode_compatible
+class TableMetadata(db.Model):
```
Should this maybe use the `TimestampMixin` to be able to track when the metadata was created and updated?

jezdez commented on redash/models/__init__.py:
```diff
@@ -63,6 +63,56 @@ def get(self, query_id):
+        }
+
+@python_2_unicode_compatible
+class ColumnMetadata(db.Model):
```
Same question about using `TimestampMixin` here?

jezdez commented on redash/models/__init__.py:
```diff
@@ -137,28 +187,37 @@ def get_by_id(cls, _id):
     def delete(self):
+        # Delete the relevant metadata about a data source first.
+        tables = TableMetadata.query.filter(TableMetadata.data_source_id == self.id).options(load_only('id'))
+        ColumnMetadata.query.filter(ColumnMetadata.table_id.in_(tables.subquery())).delete(synchronize_session=False)
```
```suggestion
        ColumnMetadata.query.filter(
            ColumnMetadata.table_id.in_(tables.subquery())
        ).delete(synchronize_session=False)
```

jezdez commented on redash/query_runner/__init__.py:
```diff
@@ -118,6 +119,23 @@ def _run_query_internal(self, query):
+        if error is not None:
+            raise Exception("Failed getting table sample.")
+
+        sample = {}
```
Let's move the default sample dictionary down into the else clause of the `if len(rows) > 0` condition.

jezdez commented on redash/query_runner/pg.py:
```diff
@@ -92,17 +93,24 @@ def _get_definitions(self, schema, query):
+        table_samples = {}
 
-        for row in results['rows']:
+        for i, row in enumerate(results['rows']):
```
This doesn't need the enumerate.

jezdez commented on redash/tasks/queries.py:
```diff
@@ -379,7 +380,64 @@ def refresh_schema(data_source_id):
+                models.db.session.query(TableMetadata).filter(TableMetadata.id==persisted_table.id).update({"table_exists": True})
+            else:
+                metadata = 'metadata' in table
+                persisted_table = TableMetadata(table_name=table_name, data_source_id=ds.id, column_metadata=metadata)
```
```suggestion
                persisted_table = TableMetadata(
                    table_name=table_name,
                    data_source_id=ds.id,
                    column_metadata=metadata,
                )
```

jezdez commented on redash/tasks/queries.py:
```diff
@@ -379,7 +380,64 @@ def refresh_schema(data_source_id):
+                    column_example = str(table['metadata'][i]['sample'])
+                    column_metadata['column_example'] = column_example
+                    if column_example and len(column_example) > 4000:
+                        column_metadata['column_example'] = column_example[:4000] + '...'
```
This could lead to decoding issues if the column example contains unicode characters. It'd be better to do:
```suggestion
                        column_metadata['column_example'] = u'{}...'.format('column_example[:4000])```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+
+        self.assertEqual(len(table_metadata), 1)
+        self.assertEqual(len(column_metadata), 1)
+        self.assertEqual(table_metadata[0].to_dict()['table_exists'], True)
```
```suggestion
        self.assertTrue(table_metadata[0].to_dict()['table_exists'])
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+
+        self.assertEqual(len(table_metadata), 1)
+        self.assertEqual(len(column_metadata), 1)
+        self.assertEqual(table_metadata[0].to_dict()['table_exists'], False)
```
```suggestion
        self.assertFalse(table_metadata[0].to_dict()['table_exists'])
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+        self.patched_get_schema.return_value = self.default_schema_return_value
+        refresh_schema(self.factory.data_source.id)
+        table_metadata = models.db.session.query(TableMetadata).all()
+        self.assertEqual(table_metadata[0].to_dict()['table_exists'], True)
```
```suggestion
        self.assertTrue(table_metadata[0].to_dict()['table_exists'])
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+        refresh_schema(self.factory.data_source.id)
+        column_metadata = models.db.session.query(ColumnMetadata).all()
+
+        self.assertEqual(column_metadata[0].to_dict()['column_exists'], True)
```
```suggestion
        self.assertTrue(column_metadata[0].to_dict()['column_exists'])
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+        column_metadata = models.db.session.query(ColumnMetadata).all()
+        self.assertEqual(len(column_metadata), 2)
+
+        self.assertEqual(column_metadata[1].to_dict()['column_exists'], False)
```
```suggestion
        self.assertFalse(column_metadata[1].to_dict()['column_exists'])
```

jezdez commented on tests/tasks/test_refresh_schemas.py:
```diff
@@ -25,3 +58,89 @@ def test_skips_paused_data_sources(self):
+        self.assertEqual(len(column_metadata), 2)
+
+        self.assertEqual(column_metadata[1].to_dict()['column_exists'], False)
+        self.assertEqual(column_metadata[0].to_dict()['column_exists'], True)
```
```suggestion
        self.assertTrue(column_metadata[0].to_dict()['column_exists'])
```

jezdez commented on redash/models/__init__.py:
```diff
@@ -64,7 +65,7 @@ def get(self, query_id):
 @python_2_unicode_compatible
-class TableMetadata(db.Model):
+class TableMetadata(TimestampMixin, db.Model):
     id = Column(db.Integer, primary_key=True)
```
@emtwo This will require an update to the migration file.

emtwo replied:
oops! Thank you! I've updated the migration file

jezdez commented:
@emtwo Something about the Alembic migration isn't right, in the frontend tests it fails with:

```
alembic.util.exc.CommandError: Multiple heads are present; please specify a single target revision
```

https://circleci.com/gh/getredash/redash/14809

jezdez commented:
@emtwo Thank you!

arikfr commented on redash/models/__init__.py:
```diff
@@ -138,28 +191,40 @@ def get_by_id(cls, _id):
+        tables = TableMetadata.query.filter(TableMetadata.data_source_id == self.id).options(load_only('id'))
+        ColumnMetadata.query.filter(
+            ColumnMetadata.table_id.in_(tables.subquery())
+        ).delete(synchronize_session=False)
```
Won't cascaded delete be enough here?

emtwo replied:
Yes it is! Thanks for pointing that out. Updated.

arikfr commented on redash/models/__init__.py:
```diff
@@ -64,6 +65,58 @@ def get(self, query_id):
 
 
+@python_2_unicode_compatible
+class TableMetadata(TimestampMixin, db.Model):
```
Both this and the columns table needs to include `org_id` reference.

emtwo replied:
Added, though I'm not sure if it needs to be used in a specific way or do you mean in case it needs to be referenced in the future?

arikfr commented on redash/models/__init__.py:
```diff
@@ -138,28 +191,40 @@ def get_by_id(cls, _id):
         db.session.commit()
         return res
 
     def get_schema(self, refresh=False):
```
What about the refresh feature? Users should be able to trigger schema refresh.

emtwo replied:
Indeed, I had taken it out since refreshing with samples is slower and I didn't think it was a good idea to do it synchronously. However, I've added back the ability to refresh.

arikfr commented on redash/query_runner/athena.py:
```diff
@@ -153,11 +154,20 @@ def get_schema(self, get_stats=False):
             if table_name not in schema:
-                schema[table_name] = {'name': table_name, 'columns': []}
+                schema[table_name] = {'name': table_name, 'columns': [], 'metadata': []}
+                table_samples[table_name] = self._get_table_sample(table_name)
```
Some users might have hundreds if not thousands of tables. We should probably:

1. Make loading of table samples optional.
2. Separate it from loading the schema: first load the schema, then trigger sample data loading. Each table should be in its own task.

Also, not sure we should load samples on every schema refresh. Although this is an optimization we can perform later.

emtwo replied:
I added a checkbox "Show Data Samples" in the data source for an admin to decide if these samples should be computed. They're also now computed in a separate task per table. 

In the next PR when the admin has controls per-table, we can make it possible to decide per table if samples are computed instead of per data source.

arikfr commented on redash/models/__init__.py:
```diff
@@ -138,28 +191,40 @@ def get_by_id(cls, _id):
+        for table in tables:
+            table_info = {
+                'name': table.table_name,
+                'exists': table.table_exists,
```
We should probably show only existing tables/columns in the UI. We persist them in the DB, so no data is lost, but I don't see much value in showing them to the user.

emtwo replied:
They were being filtered on the front-end initially but now they get filtered out in the backend

arikfr commented on redash/models/__init__.py:
```diff
@@ -64,6 +65,58 @@ def get(self, query_id):
+class TableMetadata(TimestampMixin, db.Model):
+    id = Column(db.Integer, primary_key=True)
+    data_source_id = Column(db.Integer, db.ForeignKey("data_sources.id"))
+    table_exists = Column(db.Boolean, default=True)
```
We store a boolean of whether the object exists, but we need some indication on when we can delete them. We could use `updated_at` as the time it was detected as missing, but for this to work we need to make sure the records are not touched after being marked as exists=False.

emtwo replied:
I made this adjustment. Now records set to exists=False do not get updated and they set the `updated_at` timestamp which is periodically checked in a celery task to prune old data

arikfr commented on redash/models/__init__.py:
```diff
@@ -64,6 +65,58 @@ def get(self, query_id):
+    data_source_id = Column(db.Integer, db.ForeignKey("data_sources.id"))
+    table_exists = Column(db.Boolean, default=True)
+    table_name = Column(db.String(255))
+    table_description = Column(db.String(4096), nullable=True)
```
This is a table object, so I guess the `table_` prefix for columns is a bit redundant? (same goes for `column_` in the Column model).

emtwo commented:
Here's what the new schema response from query runner looks like. Basically we keep the `columns` field and add an optional additional `metadata` field):
```
{
  <table_name1>: {
    'columns': [<col1_name>, <col2_name> ...],
    'metadata': [{
      "name": <col1_name>,
      "type": <col1_type>
    }, ...]
  },
  <table_name2>: { ... },
  ...
}
```

Here are a couple of screenshots. (note the new "Show data samples" option in the data source page.)
<img width="1679" alt="screen shot 2019-02-27 at 1 51 01 pm" src="https://user-images.githubusercontent.com/784781/53514973-f3205a80-3a96-11e9-9e85-d55468de65d3.png">
<img width="1142" alt="screen shot 2019-02-27 at 1 51 30 pm" src="https://user-images.githubusercontent.com/784781/53514983-f61b4b00-3a96-11e9-83a2-eddb6f9474d3.png">

emtwo commented:
@arikfr I've addressed your comments and the PR is ready for a final lookover.

emtwo commented:
Rebased on master, including this PR: https://github.com/getredash/redash/pull/3534

jezdez commented on redash/handlers/data_sources.py:
```diff
@@ -52,6 +53,9 @@ def post(self, data_source_id):
         models.db.session.add(data_source)
 
+        # Refresh the stored schemas when a data source is updated
+        refresh_schemas.apply_async(queue="schemas")
```
@emtwo Hey, after #3534 this and the other individual task calls should use the `SCHEMAS_REFRESH_QUEUE` setting now.

emtwo replied:
thanks @jezdez, I've updated the PR.

arikfr commented:
Looks like there is no re-open button? I guess a new PR will be needed.

Btw, the migrations has a bug. The foreign key references of `org_id` needs to reference `organizations.id` and not `organizations.id.id`.

jezdez commented:
Yeah, @emtwo would you mind opening a new PR please?
